Add allowedBlocks field to block.json to specify allowed children

jsnajdr opened the pull request:
**Update:** The name of the field is now `allowedBlocks`, not `children`, and the option on `useInnerBlocksProps` is not deprecated, merely reserved for specialized, dynamic usage. See https://github.com/WordPress/gutenberg/pull/58262#issuecomment-1914339111 for details.

`block.json` now supports new field, `children`, with a list of blocks that can be the children of this one. Similar to existing `parent` and `ancestor` fields where a block can limit who its parents and ancestors can be.

I'm not adding any `__experimental` prefix because in my opinion there is nothing experimental about this new field. It's a fairly obvious addition, and `parent` and `ancestor` are a good prior art.

It replaces the `allowedBlocks` option that can be passed to `useInnerBlocksProps`. This option is passed further down, only to be eventually used by `useNestedSettingsUpdate` which writes it to the `block-editor` state accessible with `.getBlockListSettings()`.

Then the only place where `.getBlockListSettings().allowedBlocks` is ever read is the `canInsertBlockType` selector.

This `allowedBlocks` functionality will still be there, for backward compatibility, but if the block type defines `children`, only this list will be used to determine valid children. `allowedBlocks` is used only as a fallback when the `children` field is not there.

I migrated the `core/list` and `core/navigation` blocks. If folks like what this PR does, I'll migrate all the rest.

The major advantage of this new approach is that the list of allowed children is part of the block type registration, not only of the block's React edit UI. And it can be extended with hooks. A 3rd party block can make itself a valid list item:
```js
addFilter( 'blocks.registerBlockType', 'my-list-item', ( settings, name ) => {
  if ( name === 'core/list' ) {
    return {
      ...settings,
      children: [ ...settings.children ?? [], 'my/list-item' ],
    }
  }
  return settings;
} );
```

jsnajdr committed:
Add support for children field in block.json

jsnajdr committed:
canInsertBlockType: check the parentBlockType.children allowlist for insertable blocks

jsnajdr committed:
Migrate core/list to .children

jsnajdr committed:
Migrate core/navigation to .children

skorasaurus commented:
This feature has been requested at https://github.com/WordPress/gutenberg/issues/16560

tjcafferkey commented on schemas/json/block.json:
```diff
@@ -73,6 +73,13 @@
 				"type": "string"
 			}
 		},
+		"children": {
```
Might be nit-picking but I feel like a more descriptive name would be better. E.g. `validChildren` or `allowedChildren` would indicate its a property imposing limitations/constraints on the block.

I was just thinking about people using `.children`, I can easily see people thinking "_Are these my blocks actual children, or allowed children?_".

I don't feel super strongly about this so if you think it's fine as-is I'm open to that.

jsnajdr replied:
I wanted to be consistent with the existing `parent` and `ancestor` fields. They aren't named `validParents` or `allowedAncestors` either. That's why I chose the plain `children`.

tjcafferkey replied:
Yeah, makes sense. Like I said, was a nit-pick above anything else so fine with me. 👍🏻

tjcafferkey reviewed (approved):
This tests well for me and besides my comment https://github.com/WordPress/gutenberg/pull/58262#discussion_r1467350870 which has been addressed, this change LGTM.

Worth noting that even though it's not a regression, using the Block Hooks API you're still able to insert non-allowed children into blocks (e.g. the Navigation block). This is fine as its how it works currently, but I'm raising incase this is something we want to think about/address in the future.

jsnajdr commented:
> using the Block Hooks API you're still able to insert non-allowed children into blocks (e.g. the Navigation block).

Yes, that's probably because the Block Hooks API runs on the server, and the block parser (`wp.blocks.parse`) doesn't check the block structure: it will parse any well-formed block markup. Maybe we should add these checks to the block parser validation step.

jsnajdr committed:
Update documentation

nerrad commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1565,14 +1565,25 @@ const canInsertBlockTypeUnmemoized = (
 		blockName
 	);
 
+	// If decision can't be made with `blockType.children` (i.e., it's not there)
+	// decide using the `allowedBlocks` block list setting (as a legacy fallback).
+	if ( hasParentAllowedBlock === null ) {
+		const parentAllowedBlocks = parentBlockListSettings?.allowedBlocks;
+		hasParentAllowedBlock = checkAllowList(
+			parentAllowedBlocks,
+			blockName
+		);
+	}
+
```
I appreciate you've added this fallback (and the rationale for the name change), yet I still wonder if the value of the name change outweighs the additional code that would need to remain indefinitely to cover the backward compatibility. There's also the long tail of documentation out there that is referencing `allowedBlocks`.

jsnajdr replied:
I think the name change is not related to this. Until now the `allowedBlocks` were a part of the block's `edit` UI, passed as a prop to the `<InnerBlocks>` component. Example:
```js
function ColumnsEdit() {
  const blockProps = useBlockProps();
  const props = useInnerBlocksProps( blockProps, { allowedBlocks: [ 'core/column' ] } );
  return <div { ...props } />;
}
```
This way the `allowedBlocks` info is buried inside React code and is not extensible.

The core idea of this PR is to move this info to `block.json`. Currently I'm adding it as a `children` field, but even if it was `allowedBlocks`, it remains true that we must continue to support `InnerBlocks.allowedBlocks`, and that there is a long tail of documentation that references `InnerBlocks.allowedBlocks`.

tyxla replied:
How confusing would it be if we rename `children` to `allowedBlocks` and lean on the idea proposed in that PR that if `allowedBlocks` is not defined in block.json, the prop passed to `InnerBlocks` will be used, but if it's defined, it would take precedence?

tellthemachines replied:
I'd support using `allowedBlocks` instead of `children`. It's immediately clear what it means, and I don't think it's confusing at all, given that we already have the concept of `allowedBlocks`. We're just moving the place where they're defined. They even have the same shape!

ockham commented:
> > using the Block Hooks API you're still able to insert non-allowed children into blocks (e.g. the Navigation block).
> 
> Yes, that's probably because the Block Hooks API runs on the server, and the block parser (`wp.blocks.parse`) doesn't check the block structure: it will parse any well-formed block markup. Maybe we should add these checks to the block parser validation step.

My preferred course of action would be to modify Block Hooks to check if a hooked block is allowed by it's designated parent block's `children` field (if any), and only insert it if it is. (We even have an issue for that: #53989.)
Not a blocker to this PR IMO; it's something we can tackle separately (and will need to anyway, as the Block Hooks code lives in Core).

gziolo commented:
Overall, I like the direction and appreciate the care for introducing an API similar to `parent` and `ancestor`. It looks like the final name is the only remaining decision.

 On the implementation side, there is one edge case that needs to be taken into account. It's an ongoing effort that was started to allow serializing the list of `allowedBlocks` as an attribute so it helps to narrow down what's possible with patterns and in similar cases. Related issue: https://github.com/WordPress/gutenberg/issues/15682. Example PR: https://github.com/WordPress/gutenberg/pull/49981.

The outcome of that is that whenever the block serializes `allowedBlocks` like in:

```html
<!-- wp:media-text {"mediaId":8,"mediaLink":"http://localhost:8888/?attachment_id=8","mediaType":"image","allowedBlocks":["core/paragraph"]} -->
<div class="wp-block-media-text is-stacked-on-mobile"><figure class="wp-block-media-text__media"><img src="http://localhost:8888/wp-content/uploads/2023/04/ros-buehne-1001900x400-1-1024x216.jpg" alt="" class="wp-image-8 size-full"/></figure><div class="wp-block-media-text__content"></div></div>
<!-- /wp:media-text -->
```

The expectation would be that `"core/paragraph"` would be the only allowed child block. So it feels like it should take precedence over the setting that the block type provides with `children`. The remaining question is whether, the block attribute should allow any block or only subset of the allowed blocks.

jsnajdr committed:
Migrate remaining Core blocks

jsnajdr commented on packages/block-library/src/form-submit-button/edit.js:
```diff
@@ -23,7 +23,6 @@ const TEMPLATE = [
 const Edit = () => {
 	const blockProps = useBlockProps();
 	const innerBlocksProps = useInnerBlocksProps( blockProps, {
-		allowedBlocks: TEMPLATE,
```
This was not a valid value for `allowedBlocks`, it led to the list being effectively empty and rejecting all child blocks. Added in #44214. @aristath can you confirm that it's correct how I converted the list to `[ 'core/buttons', 'core/button' ]`?

aristath replied:
Confirmed :+1:

jsnajdr commented:
The existence of the `allowedBlocks` attribute on some blocks (Cover, Group, ...) is a very good reason to call the `block.json` field also `allowedBlocks`. Then we have three variants:
- in `block.json`. Determined by block developer, extensible by plugins, non-overridable by content authors.
- as a block attribute: intended for content authors, when they want to create a template and then lock it down, limiting what the template users can edit.
- in `InnerBlocks`: legacy, shouldn't be used by block developer. Except when implementing the `allowedBlocks` block attribute 🙂 Then the `Edit` function reads `allowedBlocks` from `attributes` and passes it down to `useInnerBlocksProps`.

jsnajdr commented on packages/block-library/src/social-links/edit.native.js:
```diff
@@ -17,11 +17,6 @@ import { compose, usePreferredColorSchemeStyle } from '@wordpress/compose';
-
-const ALLOWED_BLOCKS = variations.map(
-	( v ) => `core/social-link-${ v.name }`
-);
```
I'm not sure about this. The web version of the Social Links component specifies allowed children as `[ 'core/social-link' ]`. But the mobile version enumerates all the variations! That looks wrong, not only because the web version doesn't do it, but also because a plugin can register a custom variation. But Social Links won't know about it, because it only knows the hardcoded list of Core variations.

jsnajdr commented:
The `core/navigation-submenu` block does something with `allowedBlocks` that can't be converted to a `block.json` field. It filters the `ALLOWED_BLOCKS` like this:
```js
const allowedBlocks =
  parentCount >= maxNestingLevel
    ? ALLOWED_BLOCKS.filter(
      ( blockName ) => blockName !== 'core/navigation-submenu'
    )
    : ALLOWED_BLOCKS;
```
This way, a limit on number of nested navigation submenus is implemented.

noisysocks commented:
> in InnerBlocks: legacy, shouldn't be used by block developers. Except when implementing the allowedBlocks block attribute 🙂 Then the Edit function reads allowedBlocks from attributes and passes it down to useInnerBlocksProps.

Could this be done automatically so that we can actually `deprecate()` `allowedBlocks` in `InnerBlocks`?

youknowriad commented:
Did we measure any performance impact. I know that `useNestedSettingsUpdate` is responsible of a chunk of the "firstBlockLoad" metric and making these settings static (part of the registration) is hopefully going to improve performance as well.

jsnajdr commented:
> I know that `useNestedSettingsUpdate` is responsible of a chunk of the "firstBlockLoad" metric

I don't see any significant changes on the performance CI job. And all the nested settings updates are happening, I just removed one `allowedBlocks` field from the calls. If we want to hope for a performance improvement, we should change `useNestedSettingsUpdate` so that it checks for the case where all the arguments are `undefined` and doesn't dispatch the `updateBlockListSettings` action in that case.

For example, there are blocks that call `useInnerBlocksProps` only with the `template` and `renderAppender` options, and none of them are part of block list settings, we're syncing a dummy object there.

jsnajdr committed:
allowedBlocks list setting is not deprecated, can be used for dynamic behavior

jsnajdr committed:
Rename children to allowedBlocks

jsnajdr commented:
Thanks everyone for your interest and nice feedback in this PR 🙂 

I made some changes:
1. Changed the `children` name to `allowedBlocks`. In the beginning I didn't know that there is a popular pattern to specify `allowedBlocks` also a block attribute. Now there are three places where you can specify `allowedBlocks`: `block.json`, block attribute, or the `useInnerBlocksProps` option. It makes sense to make all the three names consistent.
2. I'm no longer marking the `useInnerBlocksProps` option as deprecated. It can be used to make the list dynamic, different for each block. Like when implementing the `allowedBlocks` attribute, or limiting the level of nesting. That's a valid use case. The `block.json` way is now merely the preferred one, if all you need is a static list.

The `block.json` field and the `useInnerBlocksProps` option interact in such a way that the `block.json` is authoritative, individual blocks cannot expand it. If `block.json` doesn't allow a block, it can't be there. `useInnerBlocksProps` can only disallow more blocks.

I believe I addressed all feedback by now, the PR is ready for testing and merging.

jsnajdr committed:
Rename children to allowedBlocks in registration

gziolo commented:
@jsnajdr, we will need to add support for `allowedBlocks` on the server when scanning `block.json` in `register_block_type_from_metadata`, on `WP_Block_Type` and expose it in REST API endpoint. There is still also this special function `get_block_editor_server_block_settings` that exposes the registered block types to the client that needs an update.

jsnajdr committed:
canInsertBlockType: never downgrade result from boolean to null

Mamaduka commented on packages/blocks/src/store/reducer.js:
```diff
@@ -92,6 +92,20 @@ function bootstrappedBlockTypes( state = {}, action ) {
 					};
 				}
+
+				// The `children` prop is not yet included in the server provided
+				// definitions and needs to be polyfilled. This can be removed when the
+				// minimum supported WordPress is >= 6.5.
+				if (
+					serverDefinition.children === undefined &&
+					blockType.children
+				) {
+					newDefinition = {
+						...serverDefinition,
+						...newDefinition,
+						children: blockType.children,
+					};
+				}
```
We need to update names in this condition as well.

jsnajdr replied:
Both places updated now 🙂

jsnajdr committed:
Rename children to allowedBlocks also when patching server definitions

jsnajdr committed:
Add allowed_blocks to gutenberg_add_hooked_blocks

jsnajdr commented on lib/compat/wordpress-6.4/block-hooks.php:
```diff
@@ -73,6 +73,7 @@ function gutenberg_add_hooked_blocks( $settings, $metadata ) {
 		'keywords'         => 'keywords',
 		'example'          => 'example',
 		'variations'       => 'variations',
+		'allowed_blocks'   => 'allowedBlocks',
```
@ockham I added the `allowed_blocks` field also here, but I don't really know what I'm doing. Is the field relevant in the block hooks code?

ockham replied:
I think it's fine to add here. This is the compatibility layer that implements the Block Hooks "shim" for when the GB plugin runs on WP < 6.4; we're [adding the `blockHooks` field](https://github.com/WordPress/gutenberg/pull/58262/files#diff-0019d64feee50979366751fa8f69fda30459ebfa3e5fc2b15ee295feae284c78R79) to the preloaded block settings (which unfortunately involves duplicating [the `$fields_to_pick` allowlist from Core](https://github.com/WordPress/wordpress-develop/blob/2ce87aa6f530dc3a80780f8c3b65f5f17d3d1321/src/wp-admin/includes/post.php#L2287-L2306), as it's not exposed by any filter or the like).

Since we're now also adding the `allowed_blocks` field in `block.json`, it makes sense to include that field in the allowlist 👍

ntsekouras replied:
I don't think we should update files under `6.4` compat folder. If we intend to use this feature in `6.5`, we should have a new file under that folder and I guess it will require a backport.  --cc @gziolo @Mamaduka

gziolo replied:
Once we have the WordPress Core part as outlined in https://github.com/WordPress/gutenberg/pull/58262#issuecomment-1914493617, then it definitely might become an issue when support for WordPress 6.3 in the Gutenberg plugin drops and this polyfill is no longer there.

Mamaduka replied:
Nik is right. PHP code should be moved to a target WP version compat directory to avoid missing the backport.

jsnajdr replied:
Here I modified `gutenberg_add_hooked_blocks` function that's intended to polyfill the block hooks behavior for WP < 6.4. WordPress 6.4 and newer already has the relevant code directly in the `register_block_type_from_metadata` function and doesn't need any patching.

I'll need to add a separate WordPress 6.5 back compat code in order to support `allowedBlocks`, both in Core and in the Gutenberg plugin. But I haven't done that yet.

jsnajdr replied:
@gziolo @ockham https://core.trac.wordpress.org/ticket/60403 and https://github.com/WordPress/wordpress-develop/pull/5988 should be all we need in Core to support `allowedBlocks`. I closely followed what Bernie did when adding `blockHooks`.

gziolo replied:
Excellent, I’ll take it from here and commit tomorrow unless Bernie is faster. Thank you so much @jsnajdr. Stellar job!

ockham commented on docs/how-to-guides/block-tutorial/nested-blocks-inner-blocks.md:
```diff
@@ -109,12 +111,13 @@ add_action( 'init', function() {
+## Using parent, ancestor and children relationships in blocks
 
-A common pattern for using InnerBlocks is to create a custom block that will be only be available if its parent block is inserted. This allows builders to establish a relationship between blocks, while limiting a nested block's discoverability. Currently, there are two relationships builders can use: `parent` and `ancestor`. The differences are:
+A common pattern for using InnerBlocks is to create a custom block that will be only be available if its parent block is inserted. This allows builders to establish a relationship between blocks, while limiting a nested block's discoverability. There are three relationships that builders can use: `parent`, `ancestor` and `allowedBlocks`. The differences are:
```
```suggestion
A common pattern for using InnerBlocks is to create a custom block that will only be available if its parent block is inserted. This allows builders to establish a relationship between blocks, while limiting a nested block's discoverability. There are three relationships that builders can use: `parent`, `ancestor`, and `allowedBlocks`. The differences are:
```

ockham reviewed (approved):
I left a [minor note](https://github.com/WordPress/gutenberg/pull/58262#pullrequestreview-1848702354), but otherwise, LGTM! :shipit: 

Thanks a lot for implementing this, @jsnajdr!

jsnajdr committed:
Fix typo

ockham commented:
It might make sense to include this in GB 17.6 so it gets a bit of exposure before we land it in Core for WP 6.5.
cc/ @c4rl0sbr4v0 😊

c4rl0sbr4v0 commented:
And I will highlight it on the Make Core post, for sure!

c4rl0sbr4v0 commented:
I just cherry-picked this PR to the release/17.6 branch to get it included in the next release: 9bb6d3e85f

getdave commented:
👋 I assume this is targeted for WP 6.5?

Can I confirm these are the correct references?

- https://core.trac.wordpress.org/ticket/60403
- https://github.com/WordPress/wordpress-develop/pull/5988

getdave commented:
Is there a way to add a block to the allowedBlocks list via PHP filters?

gziolo commented:
Yes, it should be possible with [`register_block_type_args'](https://developer.wordpress.org/reference/hooks/register_block_type_args/).

AndrewJMcDonald commented:
Are there plans to add an `allowedBlocks` property to block variations ? That would be so incredibly useful.

getdave commented:
> That would be so incredibly useful.

Could you describe a likely use case for this?

jsnajdr commented:
For block variations, a good approach could be supporting the `allowedBlocks` attribute and setting it on variations. After all, a block variation is a "block type + certain set of attributes".

AndrewJMcDonald commented:
@getdave 

Lets imagine I were building my client a columns block which is simpler than the `core/columns` block because it doesn't have features such as column widths, vertical alignment, column spacing etc. I'll take care of the styling for them.
My columns block (`my/columns`) might allow for  `my/column` and `core/image` blocks, except when `my/columns` as a four-column variant, in which case my design might have a requirement that I allow `my/column` blocks only and not images. 

I can currently achieve this by making separate blocks ie `my/columns`, and  `my/four-columns` which have different `allowedBlocks`, but it would save me a lot of time of i could set the allowed blocks per variation.

draganescu commented:
hi @AndrewJMcDonald - interesting use case! I would think maybe a combination of [block locking](https://developer.wordpress.org/block-editor/how-to-guides/curating-the-editor-experience/block-locking/) and [block templates](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-templates/) would help more.

What I am saying is allowed blocks should not be used except for limiting children that the parent can't deal with. In your example, your columns block can deal with images fine, but one variation _should not contain them for styling reasons_ and that is not a variation as much as it's a template limitation. You could for instance define a runtime in your columns block that sets a locked template for inner blocks, with a different `allowedBlocks` for the inner blocks if the number of columns is more than 3.

jsnajdr commented:
@AndrewJMcDonald Here's one way how to do this. In `block.json`, you define a "baseline": all `my/columns` blocks allow only columns and images, nothing else:
```js
{
  name: 'my/columns',
  allowedBlocks: [ 'my/column', 'core/image' ]
}
```
Then, in the block's edit function, you can limit it further:
```js
function MyColumnsEdit( { name, attributes } ) {
  const variation = useSelect( ( select ) => select( blocksStore ).getActiveBlockVariation( name, attributes ) );
  let allowedBlocks = undefined;
  if ( variation === 'four-columns' ) {
    allowedBlocks = [ 'my/column' ];
  }
  const props = useInnerBlocksProps( {}, { allowedBlocks } );
  return <div { ...props } />;
}
```
The key points here are:
1. Block variation is determined by a matching set of block attributes, nothing else. Variations define an `isActive` function that looks at the set of `attributes` and returns a boolean value that says "yes this is me" or "no not me". That's what the `getActiveBlockVariation` call does. It determines whether we're dealing with the `four-columns` variation or not.
2. If the variation is `four-columns`, create a custom list of `allowedBlocks` and pass it to `useInnerBlocksProps`.
3. The hook returns props that are then passed to an element like `div`. One of the props is `children`, and these children contain an `<InnerBlocks>` component instance that can limit its `allowedBlocks`, in addition to what is in the `block.json`.

You can find some very similar prior art in the `core/navigation-submenu` block. This block can contain a nested `core/navigation-submenu` and a few other block types. But the nested submenu is not allowed if the nesting level is too deep. The block uses the same technique as in my example above: from the dynamic context (attributes, parent blocks) it determines whether we're dealing with a particular variation or pseudo-variation, and then dynamically determines `allowedBlocks` for its inner blocks.

I'm not familiar with the block locking and block templates that @draganescu recommends. Maybe they work equally well or even better!

Mamaduka commented:
@jsnajdr, I noticed that `allowedBlocks` from the block type definition isn't synchronized with block list settings (`getBlockListSettings( rootClientId)`). Was this omitted intentionally, or was it just an oversight?

P.S. Happy to create a follow-up PR in case of the latter.

jsnajdr commented:
> Was this omitted intentionally, or was it just an oversight?

I'd say it was somewhat intentional 🙂 The `allowedBlocks` definition is used only by the `canInsertBlockType` selector. And the block list settings serve only as a "delivery mechanism" how to pass that value to the `canInsertBlockType` function.

When the `allowedBlocks` value is dynamically calculated by the block's edit function, then the block list settings are a good location where to put the value into the `block-editor` Redux state. But the block type definition value can be read directly, it's already in the `blocks` store.

I'm wondering what would be the use case for having block type `allowedBlocks` in the block list settings?

jsnajdr commented:
## Dev note

## Support for new `allowedBlocks` field in `block.json`

WordPress 6.5 adds support for new field `allowedBlocks` in the `block.json` file. It lets block developers specify which block types can be inserted as children of the given block. It's a companion to the existing `parent` and `ancestor` fields that have a similar function, namely specifying allowed parent block types. Example of usage in a `block.json` file:
```json
{
  "name": "core/list",
  "title": "List",
  "allowedBlocks": [ "core/list-item" ]
}
```
One of the main advantages of the `block.json` field is that it can be modified by plugins. For example, you can create a block that acts as a custom list item, and extend the Core List block so that your custom block can be inserted as its child:
```js
addFilter( 'blocks.registerBlockType', 'my-list-item', ( settings, name ) => {
  if ( name === 'core/list' ) {
    const { allowedBlocks = [] } = settings;
    return { ...settings, allowedBlocks: [ ...allowedBlocks, 'my/list-item' ] };
  }
  return settings;
} );
```

This new API replaces the `allowedBlocks` option that can be passed to the `useInnerBlocksProps` hook inside the block's `edit` function. This option was previously used to implement the same check, but it wasn't extensible: the list of allowed blocks was hardcoded in the block's `edit` function and couldn't be modified.

The `allowedBlocks` option on the `useInnerBlocksProps` hook is still supported and is not deprecated, but you should use it only for specialized use cases. Like when the list of allowed blocks is dynamically calculated for each block: from its attributes, or from its surrounding environment (the parent block etc.).
